bin/pasta_file_change_recipients.py

```diff
@@ -0,0 +1,108 @@
+
+log = getLogger(__name__[-15:])
+
+RECIPIENTS = None
+
+def dump_csv(config, f_recipients):
+    repo = config.repo
+    global RECIPIENTS
+    RECIPIENTS = defaultdict(Counter)
+    for message_id in tqdm(repo.mbox.get_ids(time_window=config.mbox_time_window)):
+        patch = repo[message_id]
+        mail = repo.mbox.get_messages(message_id)[0]
+        recipients = email_get_recipients(mail)
+        for filename in patch.diff.affected:
+            for recipient in recipients:
+                RECIPIENTS[filename][recipient] += 1
+    
+    if f_recipients:
+        with open(f_recipients, 'w') as f:
+            f.write('filename, recipients, count\n')
+            for filename, ctr in RECIPIENTS.items():
+                ctr = sorted(ctr.items(), key = lambda x: x[1], reverse=True)
+                f.write(f"{filename}, {','.join(map(lambda x: x[0]+','+str(x[1]),ctr))}\n")
+                RECIPIENTS[filename] = set(map(lambda x: x[0], ctr))
+
+
+
+def file_change_recipients(config, argv):
+
+    parser = argparse.ArgumentParser(prog='predict', description='Get recipients of the patch file passed \
+                                    considering the files changed in the diff.')
+
+    parser.add_argument('--gen', dest='_generate', action="store_true",
+                        help='Evaluate the recipients relation with respect to files changed')
+    parser.add_argument('--use', dest='f_recipients', metavar='csv_path', default=None, type=str,
+                        help='Path to csv relation file if already present.')
+    parser.add_argument('--dump', dest='f_recipients', metavar='saveto', default=None, type=str,
+                        help='Path to the dump the relation as a CSV file')
+    parser.add_argument('--patch', dest='f_patch', metavar='patch', default=None, type=str,
+                        help='Path to the patch file')
+    parser.add_argument('--out', dest='f_out', metavar='outpath', type=str, default=None,
+                        help='Provide a path to output file containing the list of recipients related to \
+                        the patch')
+
+    args = parser.parse_args(argv)
+    repo = config.repo
+    repo.register_mbox(config)
+    global RECIPIENTS
+
+    if args._generate:
+        log.info('Evaluating recipients realtion with respect to files changed')
+        if args.f_recipients:
+            args.f_recipients = os.path.join(os.getcwd(), args.f_recipients)
+        dump_csv(config, args.f_recipients)
+
+    elif args.f_recipients:
+        args.f_recipients = os.path.join(os.getcwd(), args.f_recipients)
+        log.info(f"Loading recipients from {args.f_recipients}")
+        RECIPIENTS = defaultdict()
+        with open(args.f_recipients, 'r') as f:
+            lines = f.readlines()[1:]
+            for line in lines:
+                line = line.strip().split(",")
+                RECIPIENTS[line[0]] = set([ line[i].strip() for i in range(len(line)) if i%2 ])
```

Comment:
I bet we have a bug here. I saw some strange results... Better use python-pickle.

Reply:
Sure